libraries/chain/controller.cpp

```diff
@@ -2825,6 +2826,34 @@ struct controller_impl {
       }
    } FC_CAPTURE_AND_RETHROW() } /// apply_block
 
+   void create_and_send_vote_msg(const block_header_state& bhs) {
+#warning use decide_vote() for strong after it is implementd by https://github.com/AntelopeIO/leap/issues/2070
+      bool strong = true;
+
+      // A vote is created and signed by each finalizer configured on the node that
+      // in active finalizer policy
+#warning possible optimization: when setting finalizer policy, update finalizer_keys_on_the_node to indicate if it is active to save going over bhs.finalizer_policy->finalizers
+      // go over each active finalizer and check if it is configured on the node
+      for (const auto& f: bhs.finalizer_policy->finalizers) {
+         auto it = finalizer_keys_on_the_node.find( f.public_key );
+         if( it != finalizer_keys_on_the_node.end() ) {
+            const auto& private_key = it->second;
+
+            // signing can take long time, do it asynchronously
+            auto sig_fut = post_async_task(thread_pool.get_executor(),
+                                           [&]() {
+                                              const auto& digest = bhs.compute_finalizer_digest();
+                                              return private_key.sign(std::vector<uint8_t>(digest.data(), digest.data() + digest.data_size())); });
+
+            // construct the vote message
+            hs_vote_message vote{ bhs.id, strong, private_key.get_public_key(), sig_fut.get() };
```

Comment:
The `sig_fut.get()` blocks until async task is complete. You should loop through twice. Once to create a vector of futures and second time to create vote and emit.

Although in practice nodes will only have one finalizer key configured (except for our integration tests). So this optimization is not useful. Also unless we can do something else there is no point in doing the signing async and this extra post async is just added overhead.

Reply:
Why don't we construct the ` hs_vote_message` in the async_task, and `emit` the `future ` as a parameter to `self.voted_block`? `net_plugin` can do the wait for the future in one of its many threads.

Reply:
Would have to be a `shared_future`, but seem like a great idea.

Reply:
Thanks. For the first prototype, I change to do signing directly.

Reply:
The issue mentions that the digest should be signed asynchronously. If we merge this implementation it won't fully resolve the issue.

Reply:
Why `shared_future`? Because signal can only be copied?

I will ask @aheg if asynchronous signing is required for the prototype tomorrow.

Reply:
Because signals are sent to multiple connections to the signal. You will need to be able to `.get()` from multiple.